### PR TITLE
Added per-edge stats

### PR DESCRIPTION
**Pull request by jekka001**
## Pull Request description
Introduced support for reporting Downlink queue metrics as timeseries telemetry, analogous to the existing Uplink metrics. The following new telemetry keys are now available on the Edge.

These metrics reflect the state and performance of Downlink message delivery (Cloud → Edge) and can be visualized in custom dashboards for full synchronization observability.

The Uplink statistics are collected on the Edge, based on local message processing.
The following metrics are reported as timeseries:

- uplinkMsgsAdded – number of messages added to the Uplink queue.
- uplinkMsgsPushed – number of messages successfully pushed to the Cloud.
- uplinkMsgsPermanentlyFailed – number of messages that failed permanently and will not be retried.
- uplinkMsgsTmpFailed – number of temporarily failed messages (e.g. due to network issues).
- uplinkMsgsLag – number of pending messages currently in the Uplink queue (lag).

image

Downlink statistics follow the same pattern, using analogous telemetry keys:

- downlinkMsgsAdded
- downlinkMsgsPushed
- downlinkMsgsPermanentlyFailed
- downlinkMsgsTmpFailed
- downlinkMsgsLag

image

Both Uplink and Downlink metrics are exposed as standard timeseries telemetry in ThingsBoard and can be visualized in dashboards, enabling full observability of Edge ↔ Cloud synchronization.
image

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [ ] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [ ] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    private final TbKafkaSettings kafkaSettings;
+    private final TimeseriesService tsService;
+
+    @Value("${queue.kafka.lag-report.send-ts-interval-ms:60000}")
+    private long reportIntervalMs;
+
+    private KafkaConsumer<String, byte[]> consumer;
+    private ScheduledExecutorService scheduler;
+
+    @PostConstruct
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: Let's use  @AfterStartUp(order = AfterStartUp.REGULAR_SERVICE) instead of PostConstruct

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+    private void reportLag() {
+        try {
+            AdminClient adminClient = kafkaSettings.getAdminClient();
+            Set<String> groupIds = adminClient.listConsumerGroups()
+                    .all()
+                    .get(10, TimeUnit.SECONDS)
+                    .stream()
+                    .map(ConsumerGroupListing::groupId)
+                    .filter(id -> id.startsWith("tb_edge_event.notifications."))
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: Let's use combination of TB_QUEUE_PREFIX + TB_QUEUE_EDGE_EVENT_NOTIFICATIONS_TOPIC env instead of hard coded value.

Comment (volodymyr-babak) [reply application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java:85]: TB_QUEUE_EDGE_EVENT_NOTIFICATIONS_TOPIC

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                    .collect(Collectors.toSet());
+
+            log.debug("Found {} consumer groups with prefix 'tb_edge_event.notifications.'", groupIds.size());
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: the same as before

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+                ListConsumerGroupOffsetsResult offsetsResult = adminClient.listConsumerGroupOffsets(groupId);
+                Map<TopicPartition, OffsetAndMetadata> groupOffsets = offsetsResult.partitionsToOffsetAndMetadata().get(5, TimeUnit.SECONDS);
+                Map<TopicPartition, Long> endOffsets = consumer.endOffsets(groupOffsets.keySet());
+
+                for (Map.Entry<TopicPartition, OffsetAndMetadata> entry : groupOffsets.entrySet()) {
+                    TopicPartition tp = entry.getKey();
+                    String topic = tp.topic();
+                    long committed = entry.getValue().offset();
+                    Long latest = endOffsets.get(tp);
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: let's rename latest to end or endOffset to be in sync with Kafka naming conventions

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                }
+            }
+        } catch (Exception e) {
+            log.warn("Error reporting Edge Kafka lag: ", e);
+        }
+    }
+
+    /**
+     * Parse tenantId and edgeId from topic name.
+     * Expected format: tb_edge_event.notifications.<tenantId>.<edgeId>
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: please check what is expected format in case TB_QUEUE_PREFIX and TB_QUEUE_EDGE_EVENT_NOTIFICATIONS_TOPIC are set and if they have more . then expected in this code?

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            } catch (IllegalArgumentException e) {
+                log.debug("Failed to parse tenantId or edgeId from topic: {}", topic, e);
+            }
+        }
+        return Optional.empty();
+    }
+
+    @PreDestroy
+    public void shutdown() {
+        if (scheduler != null) scheduler.shutdownNow();
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: if (scheduler != null) {
   scheduler.shutdownNow();
}

Please stick to this format as we use it as default for TB code

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        if (consumer != null) consumer.close();
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: same as above

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                    long lag = latest - committed;
+
+                    extractTenantEdge(topic).ifPresent(tenantEdge -> {
+                        long now = System.currentTimeMillis();
+                        long roundedTs = now - (now % TimeUnit.MINUTES.toMillis(1));
+                        TsKvEntry kvEntry = new BasicTsKvEntry(roundedTs, new LongDataEntry(LAG_KEY, lag));
+
+                        ListenableFuture<TimeseriesSaveResult> future = tsService.save(
+                                TenantId.fromUUID(tenantEdge.getFirst()),
+                                new EdgeId(tenantEdge.getSecond()),
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: please add fromUUID for EdgeId class

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+                    long committed = entry.getValue().offset();
+                    Long latest = endOffsets.get(tp);
+                    if (latest == null) {
+                        log.warn("No end offset for topic-partition {}. Skipping.", tp);
+                        continue;
+                    }
+
+                    long lag = latest - committed;
+
+                    extractTenantEdge(topic).ifPresent(tenantEdge -> {
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: extractTenantEdge -> extractTenantIdEdgeId
tenantEdge -> tenantIdEdgeId

Please update names accordingly for clear readability of this code

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+                        long now = System.currentTimeMillis();
+                        long roundedTs = now - (now % TimeUnit.MINUTES.toMillis(1));
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: I suppose we need to round not to hardcoded 1 minute, but for reportIntervalMs?

File: application/src/main/resources/thingsboard.yml [hunk 11]
```diff
@@ -1693,6 +1693,9 @@ queue:
       print-interval-ms: "${TB_QUEUE_KAFKA_CONSUMER_STATS_MIN_PRINT_INTERVAL_MS:60000}"
       # Time to wait for the stats-loading requests to Kafka to finish
       kafka-response-timeout-ms: "${TB_QUEUE_KAFKA_CONSUMER_STATS_RESPONSE_TIMEOUT_MS:1000}"
+    lag-report:
```
Comment (volodymyr-babak) [inline application/src/main/resources/thingsboard.yml]: please add possibility to disable this kafka lag reporter by additinal env variable enabled/disabled

File: application/src/main/resources/thingsboard.yml [hunk 12]
```diff
@@ -1693,6 +1693,9 @@ queue:
...
+      # Interval for reporting downlink lag per Edge.
```
Comment (volodymyr-babak) [inline application/src/main/resources/thingsboard.yml]: Let's move this to edges: ?
so, it's going to be 
edges.kafka.lag-report-to-ts.enabled
edges.kafka.lag-report-to-ts.send-ts-interval-ms

Comment (volodymyr-babak) [reply application/src/main/resources/thingsboard.yml]: or downlink-lag-report-to-ts ?

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+                        long roundedTs = now - (now % reportIntervalMs);
+                        TsKvEntry kvEntry = new BasicTsKvEntry(roundedTs, new LongDataEntry(LAG_KEY, lag));
+
+                        ListenableFuture<TimeseriesSaveResult> future = tsService.save(
+                                TenantId.fromUUID(tenantIdEdgeId.getFirst()),
+                                EdgeId.fromUUID(tenantIdEdgeId.getSecond()),
+                                kvEntry
+                        );
+
+                        Futures.addCallback(future, new com.google.common.util.concurrent.FutureCallback<>() {
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: import?

File: application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+    @AfterStartUp(order = AfterStartUp.REGULAR_SERVICE)
+    public void init() {
+        if (!lagReportEnabled) {
+            log.info("EdgeKafkaLagReporter is disabled by configuration.");
+            return;
+        }
+        log.info("Initializing EdgeKafkaLagReporter with interval {}ms...", reportIntervalMs);
+        this.consumer = new KafkaConsumer<>(kafkaSettings.toConsumerProps(null));
+        this.scheduler = Executors.newSingleThreadScheduledExecutor();
+        this.scheduler.scheduleAtFixedRate(this::reportLag, 0, reportIntervalMs, TimeUnit.MILLISECONDS);
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/EdgeKafkaLagReporter.java]: let's delay with reportIntervalMs instead of 0 to decrease pressure on kafka service during startup

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    private TbKafkaAdmin tbKafkaAdmin;
+
+    @Value("${edges.stats.enabled:true}")
+    private boolean edgesStatsEnabled;
+    @Value("${edges.stats.ttl:30}")
+    private int edgesStatsTtlDays;
+    @Value("${edges.stats.report-interval-millis:20000}")
+    private long reportIntervalMillis;
+
+    @Scheduled(fixedDelayString = "${edges.stats.report-interval-millis:20000}")
```
Comment (AndriiLandiak) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: Do we need to call this immediately after start-up, or could we use initialDelay? Vova also wrote you similar in the last comment

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    public void reportStats() {
+        if (!edgesStatsEnabled) {
```
Comment (AndriiLandiak) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: You could annotate the class itself with @ConditionalOnProperty. Currently, it looks like you're still calculating and recording events even when stats are disabled, but those values aren’t used anywhere. If the class is optional and not present, the recorded data is never read or cleared — so what's the purpose of collecting it in the first place?

The same for EdgeStatsCounter?

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeStatsCounterService.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+import org.thingsboard.server.common.data.id.EdgeId;
+import org.thingsboard.server.common.data.id.TenantId;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+@Service
+@Slf4j
+@Getter
+public class EdgeStatsCounterService {
```
Comment (AndriiLandiak) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeStatsCounterService.java]: TbCore only? Same for EdgeCommunicationStatsService

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+    private static final String DOWNLINK_MSGS_LAG = "downlinkMsgsLag";
+
+    @Autowired
+    private TimeseriesService tsService;
+    @Autowired
+    private EdgeStatsCounterService statsCounterService;
+    @Autowired
+    private TopicService topicService;
+    @Autowired(required = false)
+    private TbKafkaAdmin tbKafkaAdmin;
```
Comment (AndriiLandiak) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: RequiredArgsConstructor? kafkaAdmin then is optional

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeStatsCounterService.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+    private final ConcurrentMap<EdgeId, MsgCounters> counterByEdge = new ConcurrentHashMap<>();
```
Comment (AndriiLandiak) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeStatsCounterService.java]: This map is never cleared. Consider using ConcurrentReferenceHashMap, or explicitly remove when the edge is deleted (or even when the grpc session is closed, that's up to your business logic)

**Review by AndriiLandiak — APPROVED** [review]

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@TbCoreComponent
+@ConditionalOnProperty(prefix = "edges.stats", name = "enabled", havingValue = "true", matchIfMissing = false)
+@RequiredArgsConstructor
+@Service
+@Slf4j
+public class EdgeCommunicationStatsService {
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: Choose more generic naming. The stats will most likely be extended in the future. EdgeStatsService will do

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+
+    private static final String DOWNLINK_MSGS_ADDED = "downlinkMsgsAdded";
+    private static final String DOWNLINK_MSGS_PUSHED = "downlinkMsgsPushed";
+    private static final String DOWNLINK_MSGS_PERMANENTLY_FAILED = "downlinkMsgsPermanentlyFailed";
+    private static final String DOWNLINK_MSGS_TMP_FAILED = "downlinkMsgsTmpFailed";
+    private static final String DOWNLINK_MSGS_LAG = "downlinkMsgsLag";
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: Move to CounterEventType enum

File: common/queue/src/main/java/org/thingsboard/server/queue/kafka/TbKafkaAdmin.java [hunk 22]
```diff
@@ -260,4 +263,46 @@ public boolean areAllTopicsEmpty(Set<String> topics) {
         }
     }
 
+    public Map<String, Long> getTotalLagForGroupsBulk(Set<String> groupIds) {
```
Comment (ViacheslavKlimov) [inline common/queue/src/main/java/org/thingsboard/server/queue/kafka/TbKafkaAdmin.java:265]: Please make this code more reusable. Introduce a method `getTotalConsumerGroupLag(String groupId)`

File: application/src/main/java/org/thingsboard/server/service/edge/stats/CounterEventType.java [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.server.service.edge.stats;
+
+public enum CounterEventType {
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/CounterEventType.java]: Maybe `EdgeStatsKey`?

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+
+
+    @Scheduled(
+            fixedDelayString = "${edges.stats.report-interval-millis:20000}",
+            initialDelayString = "${edges.stats.report-interval-millis:20000}"
+    )
+    public void reportStats() {
+        log.debug("Reporting Edge communication stats...");
+
+        long ts = (System.currentTimeMillis() / reportIntervalMillis) * reportIntervalMillis;
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: A bit hard to understand.
Refactor to (if you really need to truncate the time to report period, though I don't see much sense):
`now - (now % reportIntervalMillis)`

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeStatsCounterService.java [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+import org.thingsboard.server.queue.util.TbCoreComponent;
+
+@TbCoreComponent
+@ConditionalOnProperty(prefix = "edges.stats", name = "enabled", havingValue = "true", matchIfMissing = false)
+@Service
+@Slf4j
+@Getter
+public class EdgeStatsCounterService {
+
+    private final ConcurrentReferenceHashMap<EdgeId, MsgCounters> counterByEdge = new ConcurrentReferenceHashMap<>();
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeStatsCounterService.java]: Reason for ConcurrentReferenceHashMap? There will be issues when low on memory. Don't do `counterByEdge.get(edgeId).clear()` in the `clear` method, do `remove` from the map. Then you can use regular concurrent map.

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+
+        Map<EdgeId, MsgCounters> countersByEdge = statsCounterService.getCounterByEdge();
+        Map<EdgeId, Long> lagByEdgeId = tbKafkaAdmin.isPresent() ? getEdgeLagByEdgeId(countersByEdge) : Collections.emptyMap();
+        for (Map.Entry<EdgeId, MsgCounters> counterByEdge : countersByEdge.entrySet()) {
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: `countersByEdge.forEach((edgeId, counters) -> { ... })`

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        e -> lagByTopic.getOrDefault(e.getValue(), 0L)
+                ));
+    }
+
+    private void saveTs(TenantId tenantId, EdgeId edgeId, List<TsKvEntry> statsEntries) {
+        try {
+            tsService.save(tenantId, edgeId, statsEntries, TimeUnit.DAYS.toSeconds(edgesStatsTtlDays));
+            log.debug("Successfully saved edge time-series stats: {} for edge: {}", statsEntries, edgeId);
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: tsService.save is async. The log is a bit misleading since you don't wait for the completion. You also don't log any failures, if we failed to save the timeseries.

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+            }
+            List<TsKvEntry> statsEntries = List.of(
+                    entry(ts, DOWNLINK_MSGS_ADDED, counters.getMsgsAdded().get()),
+                    entry(ts, DOWNLINK_MSGS_PUSHED, counters.getMsgsPushed().get()),
+                    entry(ts, DOWNLINK_MSGS_PERMANENTLY_FAILED, counters.getMsgsPermanentlyFailed().get()),
+                    entry(ts, DOWNLINK_MSGS_TMP_FAILED, counters.getMsgsTmpFailed().get()),
+                    entry(ts, DOWNLINK_MSGS_LAG, counters.getMsgsLag().get())
+            );
+
+            log.trace("Reported Edge communication stats: {}", statsEntries);
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: Add tenantId, edgeId to the log

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeEventService.java [hunk 29]
```diff
@@ -40,6 +43,7 @@ public class KafkaEdgeEventService extends BaseEdgeEventService {
 
     private final TopicService topicService;
     private final TbQueueProducerProvider producerProvider;
+    private final Optional<EdgeStatsCounterService> statsCounterService;
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeEventService.java:46]: Does EdgeEventService run on Rule Engine? If so, don't you need events stats from there too?

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java [hunk 30]
```diff
@@ -0,0 +1,127 @@
...
+    private static final String DOWNLINK_MSGS_LAG = "downlinkMsgsLag";
+
+    private final TimeseriesService tsService;
+    private final EdgeStatsCounterService statsCounterService;
+    private final TopicService topicService;
+    private final Optional<TbKafkaAdmin> tbKafkaAdmin;
+
+    @Value("${edges.stats.ttl:30}")
+    private int edgesStatsTtlDays;
+    @Value("${edges.stats.report-interval-millis:20000}")
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: Can we make it not so frequent? In worst case, you will save 5 datapoints each 20 seconds for all of the edges. Maybe let's report it every 15-30 minutes, at least? Please talk to @volodymyr-babak about that

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+    private long reportIntervalMillis;
+
+
+    @Scheduled(
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: Is it possible for two instances to receive stats (events) for the same edge? Or partitioning is in place?

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeEventService.java (hunk 29 above)
Comment (jekka001) [reply application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeEventService.java:46]: No, EdgeEventService is not used in the Rule Engine. It's only used on the edge core side, so we don't need to collect event stats from the Rule Engine.

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java (hunk 31 above)
Comment (jekka001) [reply application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeCommunicationStatsService.java]: Since all instances use the same Kafka consumer group and messages are keyed by EdgeId, Kafka guarantees message ordering per key and ensures that each EdgeId is always processed by the same instance — so probably not.

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeStatsKey.java [hunk 32]
```diff
@@ -15,10 +15,16 @@
...
     DOWNLINK_MSG_PERMANENTLY_FAILED,
     DOWNLINK_MSG_TMP_FAILED,
-    DOWNLINK_MSG_ADDED
+    DOWNLINK_MSG_ADDED;
+
+    public static final String DOWNLINK_MSGS_ADDED = "downlinkMsgsAdded";
+    public static final String DOWNLINK_MSGS_PUSHED = "downlinkMsgsPushed";
+    public static final String DOWNLINK_MSGS_PERMANENTLY_FAILED = "downlinkMsgsPermanentlyFailed";
+    public static final String DOWNLINK_MSGS_TMP_FAILED = "downlinkMsgsTmpFailed";
+    public static final String DOWNLINK_MSGS_LAG = "downlinkMsgsLag";
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeStatsKey.java]: I meant to make it the field of the enum, e.g. 
```
DOWNLINK_MSG_PUSHED("downlinkMsgsPushed"),
```

File: application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeStatsService.java [hunk 33]
```diff
@@ -0,0 +1,146 @@
...
+                    log.debug("Successfully saved edge time-series stats: {} for edge: {}", statsEntries, edgeId);
+                }
+
+                @Override
+                public void onFailure(Throwable t) {
+                    log.warn("Failed to save edge time-series stats for edge: {}", edgeId, t);
+                }
+            }, MoreExecutors.directExecutor());
+        } finally {
+            statsCounterService.clear(edgeId);
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/edge/stats/EdgeStatsService.java:142]: You remove from the countersByEdge map while iterating over it.
This will fail with concurrent modification exception.

Please add tests for the Edge stats, so we are safe and calm
